### ISSUE TITLE
Custom semantic validation

### ISSUE THREAD

**Issue by maxlinc** [apigee-127/swagger-tools#140]
This is fairly closely related to #139 and is another good thing to keep in mind as you're walking through the code thinking about documentation, but I figured it deserves its own issue.

The documentation in #139 would essentially create a Swagger styleguide, but different organizations often have customized styleguides. Consider, for example, [GitHub](https://github.com/styleguide/javascript) vs [jQuery](http://contribute.jquery.org/style-guide/js/) javascript styleguides, or the many [configuration options](https://github.com/bbatsov/rubocop/blob/master/config/default.yml) for Rubocop.

I think it would be great if organizations can start with the built-in semantic rules for validating their Swagger docs, but can tweak the validation via parameters or by registering their own javascript callbacks.

A simple example - validating the operation "summary" based on OpenStack conventions:
- It should be in sentence case (e.g. "Create server", not "Create Server" or "create server")
- It should be short, perhaps even shorter than the minimum suggested by Swagger (e.g. 80 vs 120 characters)

Not something that needs to be done soon, but something to keep in mind if you need to revisit the validation code.

**Comment by whitlockjc** [comment]
I'm not against this but I have to tell you, semantic validation isn't really a style.  Either your Swagger document is valid per the documented specification or it is not, there is not any gray area.  I do think we could extend the plan behind #28 (semantic validation for vendor extensions) and figure out a way to use that same process to register style checkers.  Thoughts?

**Comment by maxlinc** [comment]
tl;dr: "API Warnings" in the current output are a gray area and are more closely related to style guides than to a strict specification. They are associated with suggestions rather than absolute requirements in the Swagger specification.

I somewhat agree, but I think the warnings that you already implemented for https://github.com/apigee-127/swagger-tools/issues/1 are really style rules. The check for "API operation summary is greater than 120 characters" is a great example.

In general I think semantic validation against the specification, a vendor extension, a custom rule, or a style guide are just variations of the same general concept. You might want to keep track of which standard you're validating but they should all share pretty much the same validation mechanism. I'll just refer to them generically as "checkable rules".

I think the most important thing is really the requirement level. The top of the Swagger specification says it uses the keywords for requirement levels defined in [RFC 2119](http://www.ietf.org/rfc/rfc2119.txt). Those keywords are:
- Absolute requirements: MUST, MUST NOT
- Suggested requirements: SHOULD, SHOULD NOT
- Optional requirements: MAY, OPTIONAL

It's pretty easy to map these keywords to "error" and "warning" level errors for a checkable rule. Since the specification contains some suggested requirements there are some "gray areas. The requirement "this field SHOULD be less than 120 characters" is a good example.

A styleguide or organizations specific standards would probably be mostly suggest requirements or "warning" level validation errors, but it is possible for them to have absolute requirements as well. The Swagger specification requires a license but does not limit the choices. A specific organization may have a set of approved licenses or even just a single license used by the entire organization. So an organization may want to enforce the rules: "The license name MUST be 'Apache 2.0'. The license url MUST be 'http://www.apache.org/licenses/LICENSE-2.0.html'"

I do think it may be useful to group validation errors by type, but mostly for summarizing the results for the users. I think tools should be more concerned with error levels than error types. If I was setting up a CI build to screen changes to a Swagger document before merging a PR, I would set it up to fail changes that introduce errors of any type, but would probably allow warnings.

So I think they should all work the same, but you might want to capture the type (basically the document a rule is associated with) so you could display a summary that looked something like this:

```
$ swagger-tools validate OS-KSS3-admin.json

Swagger Errors:

  #/paths/~1v2.0~1users~1{userId}~1OS-OS-KSS3~1credentials~1s3credentials~1{type}/get: API requires path parameter but it is not defined: type

OpenStack Conventions for Swagger Errors:

  #/license/name: License must be Apache 2.0

Enhanced Examples Extension Errors:

  #/paths/~1v2.0~1users~1{userId}~1OS-OS-KSS3~1credentials~1s3credentials~1{type}/get/x-examples[3]/request: Example does not comply with the request schema

Swagger Warnings:

  #/definitions/Pet: Definition is defined but is not used: #/definitions/Pet

3 errors and 1 warning
```

**Comment by whitlockjc** [comment]
I get your point.  Warnings are there (now) to tell you that you've either violated the Swagger Specification in some way that does not have a *real* impact on the API, like the summary length exceeding 120 characters, or to warn you that you might had made a mistake, like defining a model and not referencing it.  I don't think those are stylistic but I also can see why someone might want to impose some style checks so I'm not disagreeing with you on the benefit of having it.

**Comment by whitlockjc** [comment]
By the way, checkout the new [Swagger Validation](https://github.com/apigee-127/swagger-tools/blob/master/docs/Swagger_Validation.md) part of the documentation.  It's a work in progress but it gives you an idea of what semantic validations occur.  I'll start looking into extendible/configurable validations soon.

**Comment by whitlockjc** [comment]
@maxlinc If you look at the link above, you'll see the validations I'm manually doing.  Do you mind telling me which validations you'd like to change or have it be configurable to have their type changed?

As for fully customizable semantic validation, to me that means registering extra validators and that sounds very similar to #28.  I'd love your opinion on how you'd see this done.  I'm all for implementing this and #28 together but I would really like to know how you'd like to see the consumption side of this.

**Comment by whitlockjc** [comment]
I have moved #28 to this.  I think if we create a validation framework of sorts where people can opt-in to the validation of document paths, we can solve both this and the need for vendor extension validation.

**Comment by maxlinc** [comment]
Thanks. Everything looks good so far.

**Comment by whitlockjc** [comment]
I'm in the process of designing this right now and I was wondering how you'd want to use something like this.  How would you expect to register a validator and ensure that the validator only runs against the appropriate locations in the document, or do you see a different usage model?  I've got some ideas but of course you're the user so I'd love any feedback you have.

**Comment by whitlockjc** [comment]
What about something like this:

```js
registerValidator(someRegex, function (node, parent, root) {
  // Do your validation here
});
```

`someRegex` would be a regular expression that is matched against the JSON Pointer equivalent of the path to the node being processed.  So for the `get` operation of the `/pets` path, your `someRegex` regular expression would be applied to `#/paths/~1pets`.

The idea would be having a way to process a node, check to see if the node has any registered/matching validators and then invoke the validator(s) only if necessary.  Document node name would not be enough and so that's why I was thinking of a regular expression.  I guess we could use a glob for something simpler as well.
